README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

Socket tools for easier and faster use.


## Features
- `SocketBuffer` allows to read bytes with the given length: `readNBytes(int len)`&`read(byte[])`.
- Byte tools `SocketIn` & `SocketOut` easy to build and read socket body with the given structure.

## Getting started

## Usage

### Customize Socket Buffer
provide functions like java(`readNBytes(int len)`&`read(byte[])`) / go(`read(byte[])`)
```dart
  Socket s = await Socket.connect("localhost", 15002);
  SocketBuffer buffer = SocketBuffer();
  s.write("shit");
  s.listen((event) {
    log("${event.length}");
    buffer.add(event);
  }, onDone: () {
    buffer.done();
    s.destroy();
  }, onError: (err, stack) {
    buffer.err(err);
    s.close();
  });
  Uint8List bs = await buffer.readN(1);
  print(bs);
`````

### Byte tools

base*255 with the last byte of `255` ends the length block

- section
```
section_content_length (base-255) + section_content
```
- body:
```
section + section + ...
```


<!--
## Additional information
-->


CHANGELOG.md
## 0.1.0
- !fix add bytes mismatch index
- add debug mode in `SocketIn` & `SocketBuffer`

## 0.0.8
- make buffer `clear()` public and remove `clear()` in `_read()`

## 0.0.7
- fix large list range error

## 0.0.6
- improve sync lock in buffer
- add tests

## 0.0.4 & 0.0.5
- fix getSec read from raw

## 0.0.3
- make `SocketOut().add` private
- make `SocketOut().getLength` static

## 0.0.2
- fix: mutex lock error
- update tests

## 0.0.1

* TODO: Describe initial release.


test/sockettest.dart
import 'dart:developer';
import 'dart:io';
import 'dart:typed_data';

import 'package:ogios_sutils/buffer.dart';

Future<void> test() async {
  Socket s = await Socket.connect("localhost", 15002);
  SocketBuffer buffer = SocketBuffer();
  s.write("shit");
  s.listen((event) {
    log("${event.length}");
    buffer.add(event);
  }, onDone: () {
    buffer.done();
    s.destroy();
  }, onError: (err, stack) {
    buffer.err(err);
    s.close();
  });
  try {
    while (true) {
      Uint8List bs = await buffer.readN(1024);
      log("$bs");
    }
  } catch (err) {
    log("ERROR: $err");
  }
  log("done");
}

test/ogios_sutils_test.dart
import 'dart:convert';
import 'dart:developer';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:ogios_sutils/in.dart';
import 'package:ogios_sutils/out.dart';

Future test3() async {
  Socket s = await Socket.connect("192.168.0.108", 15001);
  SocketOut so = SocketOut();
  so.addBytes(Uint8List.fromList("fetch_byte".codeUnits));
  so.addBytes(Uint8List.fromList(utf8.encode("pPGWp")));
  await so.writeTo(s);
  SocketIn si = SocketIn(conn: s);
  // status
  Uint8List sec = await si.getSec();
  String status = utf8.decode(sec);
  log("$status");
  // filename
  sec = await si.getSec();
  String filename = utf8.decode(sec);
  log("$filename");
  // file data
  List<int> final_bytes = [];
  int total = await si.next();
  int current = 0;
  var temp = Uint8List(1024*64);
  int read;
  while (current < total) {
    read = await si.read(temp);
    final_bytes.addAll(temp.sublist(0, read));
    current += read;
  }
  log("${temp.length}");
}

void main() {
  test("socket test", () async {
    // test();
    // test2();
    await test3();
  });
}


test/transfer_fetch_test.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:ogios_sutils/in.dart';
import 'package:ogios_sutils/out.dart';

Future test3() async {
  Socket s = await Socket.connect("localhost", 15001);
  SocketOut so = SocketOut();
  so.addBytes(Uint8List.fromList("fetch".codeUnits));
  so.addBytes(Uint8List.fromList([0]));
  so.addBytes(Uint8List.fromList([10]));
  await so.writeTo(s);
  SocketIn si = SocketIn(conn: s);
  int len = await si.next();
  // print("sec length: $len");
  Uint8List sec = await si.getSec();
  // print("sec: $sec");
  // print("sec to string: ${String.fromCharCodes(sec)}");
  if (len == 1 && sec[0] == 200) {
    len = await si.next();
    // print("sec length: $len");
    sec = await si.getSec();
    // print("sec: $sec");
    // print("sec to string: ${String.fromCharCodes(sec)}");
  }
}

lib/in.dart
library ogios_sutils;

import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import './buffer.dart';

class SocketIn {
  SocketIn({required this.conn, debug = false}) {
    this.raw = SocketBuffer();
    this.raw.debug = debug;
    this.conn.listen((event) {
      this.raw.add(event);
    }, onDone: () {
      this.raw.done();
      this.conn.destroy();
    }, onError: (err) {
      this.raw.err(err);
      this.conn.close();
    });
  }
  Socket conn;
  late SocketBuffer raw;
  int readed = 0;
  int CurrSecLength = 0;

  Future<int> _main({required int index, required int t}) async {
    Uint8List b = await this.raw.readN(1);
    if (b.length != 1) {
      throw Exception("readNBytes fails to work: ${b}");
    }
    int current = b[0];
    if (current != 255) {
      t = await this._main(index: index + 1, t: t);
      num feat = pow(255, index);
      int addon = current * feat.toInt();
      t += addon;
    }
    return t;
  }

  Future<int> next() async {
    if (this.CurrSecLength < this.readed) {
      throw Exception("please read all of current section");
    }
    int total = await this._main(index: 0, t: 0);
    this.CurrSecLength = total;
    this.readed = 0;
    return total;
  }

  Future<Uint8List> getSec() async {
    if (this.readed < this.CurrSecLength) {
      Uint8List bs = Uint8List(this.CurrSecLength-this.readed);
      int readed = await this.read(bs);
      this.readed += readed;
      return bs;
    } else {
      int length = await this.next();
      Uint8List temp = Uint8List(length);
      int readlength = await this.read(temp);
      if (readlength != length) {
      }
      return temp;
    }
  }

  Future<int> read(Uint8List buf) async {
    if (this.CurrSecLength == this.readed) {
      throw Exception("no more bytes for current section");
    }
    if (buf.length <= this.CurrSecLength-this.readed) {
      int i = await this.raw.read(buf);
      this.readed += i;
      return i;
    } else {
      Uint8List temp = Uint8List(this.CurrSecLength-this.readed);
      int length = await this.raw.read(temp);
      this.readed += length;
      buf.setRange(0, temp.length, temp);
      return length;
    }
  }
}


lib/out.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

class SocketOut {
  SocketOut();
  static const int OUT_TYPE_BYTES = 1;
  static const int OUT_TYPE_READER = 2;

  List<Object> raw = [];
  List<int> types = [];

  void _add(Object raw, int t, Uint8List lenb) {
    this.raw.add(lenb);
    this.raw.add(raw);
    this.types.add(t);
  }

  static Uint8List _main(int last, int index) {
    Uint8List bytes;
    if (last >= 255) {
      int current = last % 255;
      bytes = _main((last / 255).toInt(), index + 1);
      bytes[index] = current;
    } else {
      bytes = Uint8List(index + 2);
      bytes[index] = last;
      bytes[index + 1] = 255;
    }
    return bytes;
  }

  static Uint8List getLength(int length) {
    return _main(length, 0);
  }

  void addBytes(Uint8List raw) {
    int length = raw.length;
    Uint8List content_length = getLength(length);
    this._add(raw, OUT_TYPE_BYTES, content_length);
  }

  void addReader(Stream raw, int length) {
    Uint8List content_length = getLength(length);
    this._add(raw, OUT_TYPE_READER, content_length);
  }

  Future writeTo(Socket writer) async {
    for (int i=0; i< this.raw.length; i++) {
      Object input = this.raw[i];
      if (i%2 == 0) {
        writer.add(input as Uint8List);
      } else {
        int t = this.types[((i-1)/2).toInt()];
        switch (t) {
          case OUT_TYPE_BYTES:
            writer.add(input as Uint8List);
            break;
          case OUT_TYPE_READER:
            Stream reader = input as Stream;
            Completer completer = Completer();
            reader.listen((event) {
              writer.add(event);
            }, onError: (err){
              throw err;
            }, onDone: (){
              completer.complete();
            });
            await completer.future;
            break;
          default:
            break;
        }
      }
    }
  }
}


lib/buffer.dart
library ogios_sutils;

import 'dart:developer';
import 'dart:typed_data';
import 'package:synchronized/synchronized.dart';

class SocketBuffer {
  /// Initial size of internal buffer.
  static const int _initSize = 1024;

  /// Free space threshold
  static const int _threshold = 1024 * 4;

  /// Reusable empty [Uint8List].
  ///
  /// Safe for reuse because a fixed-length empty list is immutable.
  static final _emptyList = Uint8List(0);

  /// lock
  final Lock lock = Lock();

  /// Current count of bytes written to buffer.
  int _available = 0;
  int get length => _available;
  bool get isEmpty => _available == 0;
  bool get isNotEmpty => _available != 0;

  /// Current read index.
  int _readIndex = 0;

  /// Current save index
  int _saveIndex = 0;

  /// Internal buffer accumulating bytes.
  ///
  /// Will grow as necessary
  Uint8List _buffer;

  SocketBuffer() : _buffer = _emptyList;

  void add(List<int> bytes) async {
    // sync lock with read and clear
    await lock.synchronized(() async {
      int byteCount = bytes.length;
      if (byteCount == 0) return;
      // how long the list should be to contain all bytes
      int required = _saveIndex + byteCount;
      // expand buffer list if space is not enough
      if (_buffer.length < required) {
        _grow(required);
      }
      assert(_buffer.length >= required);
      // add to buffer
      if (bytes is Uint8List) {
        _buffer.setRange(_saveIndex, required, bytes);
      } else {
        for (int i = 0; i < byteCount; i++) {
          _buffer[_saveIndex + i] = bytes[i];
        }
      }
      // update available bytes
      _saveIndex = required;
      _available += byteCount;

      // debug
      if (debug) log("Buffer add: $byteCount | Buffer available: $_available | save_index: $_saveIndex | Buffer capacity: ${_buffer.length}");
    });
  }

  void _grow(int required) {
    // We will create a list in the range of 2-4 times larger than
    // required.
    int newSize = required * 2;
    if (newSize < _initSize) {
      newSize = _initSize;
    } else {
      newSize = _pow2roundup(newSize);
    }
    var newBuffer = Uint8List(newSize);
    newBuffer.setRange(0, _buffer.length, _buffer);
    _buffer = newBuffer;
  }

  void clear() async {
    // if not pass the threshold, do nothing.
    if (_readIndex < _threshold) return;
    // else, clear data before read index. (sync lock with add and read)
    await lock.synchronized(() async {
      var length = _buffer.length - (_readIndex + 1);
      var newBuffer = Uint8List(length);
      newBuffer.setRange(0, length, Uint8List.view(_buffer.buffer, _readIndex));
      _buffer = newBuffer;
      _saveIndex -= _readIndex;
      _readIndex = 0;
    });
  }

  /// Rounds numbers <= 2^32 up to the nearest power of 2.
  static int _pow2roundup(int x) {
    assert(x > 0);
    --x;
    x |= x >> 1;
    x |= x >> 2;
    x |= x >> 4;
    x |= x >> 8;
    x |= x >> 16;
    return x + 1;
  }

  Future<Uint8List> _readN(int length) async {
    if (this._done && _available == 0) throw Exception("EOF");
    checkErr();
    if (length == 0) return _emptyList;
    Uint8List buffer = Uint8List(length);
    int read = await this._read(buffer);
    if (read == 0) throw Exception("EOF");
    return buffer.sublist(0, read);
  }

  Future<Uint8List> readN(int length) async {
    return await this._readN(length);
  }

  Future<int> read(Uint8List emptybs) async {
    return await this._read(emptybs);
  }

  Future<int> _read(Uint8List emptybs) async {
    if (this._done && _available == 0) throw Exception("EOF");
    if (emptybs.length == 0) return 0;
    int inputIndex = 0;
    int total_read = 0;
    int left = emptybs.length;

    while (left > 0 && (!this._done || _available != 0)) {
      checkErr();
      bool wait_flag = false;

      // sync with add and done
      await lock.synchronized(() async {
        if (_available == 0) {
          // if no data, wait for 10ms
          wait_flag = true;
        } else {
          // else
          if (_available <= left) {
            // if available data is not enough
            if (debug) log("(NOT ENOUGH) | left: $left | ava_before: ${_available} | ava_after: ${0} | readInd_before: ${_readIndex} | readInd_after: ${_readIndex + _available}");
            emptybs.setRange(inputIndex, inputIndex + _available,
                _buffer.sublist(_readIndex, _readIndex + _available));
            total_read += _available;
            left -= _available;
            inputIndex += _available;
            _readIndex += _available;
            _available = 0;
          } else {
            // else fill up empty list
            if(debug) log("(FILL LEFT) | left: $left | ava_before: ${_available} | ava_after: ${_available - left} | readInd_before: ${_readIndex} | readInd_after: ${_readIndex + left}");
            emptybs.setRange(inputIndex, emptybs.length,
                Uint8List.view(_buffer.buffer, _readIndex, left));
            total_read += left;
            inputIndex += left;
            _readIndex += left;
            _available -= left;
            left = 0;
          }
        }
      });
      // wait for data
      if (wait_flag) await Future.delayed(const Duration(milliseconds: 10));
    }

    // debug
    if (debug) log("return total_read: $total_read");
    return total_read;
  }

  //debug
  bool debug = false;

  bool _done = false;
  Exception? _err;

  void done() {
    this._done = true;
  }

  void err(Exception err) {
    if (!err.toString().toLowerCase().contains("connection reset by peer")) {
      this._err = err;
    }
    this._done = true;
  }

  bool isDone() {
    checkErr();
    return this._done;
  }

  void checkErr() {
    if (this._err != null) throw _err!;
  }
}


lib/main.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:ogios_sutils/buffer.dart';
import 'package:ogios_sutils/in.dart';
import 'package:ogios_sutils/out.dart';

Future<void> test() async {
  Socket s = await Socket.connect("localhost", 15002);
  SocketBuffer buffer = SocketBuffer();
  s.write("shit");
  s.listen((event) {
    print(event.length);
    buffer.add(event);
  }, onDone: () {
    buffer.done();
    s.destroy();
  }, onError: (err, stack) {
    buffer.err(err);
    s.close();
  });
  try {
    while (true) {
      Uint8List bs = await buffer.readN(1024);
      // print("getbs");
    }
  } catch (err) {
    print("ERROR: $err");
  }
  print("done");
}

void test1() {
  Uint8List a = Uint8List(1);
  print(a.sublist(0, 0));
}

Future test2_server() async {
  ServerSocket serverSocket = await ServerSocket.bind("localhost", 15002);
  await for (var socket in serverSocket) {
    print("received new conn: ${socket.remoteAddress} - ${socket.remotePort}");
    var bs = Uint8List.fromList("hi from server".codeUnits);
    SocketOut so = SocketOut();
    so.addBytes(bs);
    File f = File("/home/ogios/work/andorid/ogios_sutils/test/test.txt");
    int size = (await f.stat()).size;
    print("file size: $size - ${size == f.lengthSync()}");
    so.addReader(f.openRead(), (await f.stat()).size);
    await so.writeTo(socket);
    print("closing...");
    await serverSocket.close();
    await socket.close();
    break;
  }
  await serverSocket.close();
  print("server done.");
}

Future test2_client() async {
  Socket s = await Socket.connect("localhost", 15002);
  SocketIn si = SocketIn(conn: s);
  int len = await si.next();
  print("sec length: $len");
  Uint8List sec = await si.getSec();
  print("sec: $sec");

  len = await si.next();
  print("sec length: $len");
  sec = await si.getSec();
  print("sec: $sec");
  s.destroy();
  // await s.close();
  print("client done");
}

void test2() async {
  test2_server();
  test2_client();
}

void main() {
  // test();
  // test1();
  test2();
}
